### PR TITLE
Add support for DisplayInstallWarnings

### LINKED ISSUE #1786: [New Feature]: Add ability to display warning to user before upgrade
### Description of the new feature/enhancement

While upgrading programs, all my explorer windows closed with no warning due to WinMerge closing it as part of its upgrade path. Would it be possible to add a warning to programs known to interfere with others as part of their upgrades or installs?

### Proposed technical implementation details (optional)

Add a upgrade after user issues command to the tone of "This upgrade might close other programs as part of its upgrade, make sure you save your work first!" with a prompt Y/N before continuing and a flag to the upgrade command to ignore possible upgrade closures to disable the prompt for scripts and people who don't care

### PR DESCRIPTION
**Pull request by ryfu-msft**
Resolves #1786

Changes:
- Adds support for prompting the user with a warning message if the DisplayInstallWarnings field is set to true. The message notifies the user that the installer may interfere with running applications
- Added `--ignore-warnings` flag to allow users to override the default behavior of warning prompts

Tests:
- Added unit tests to verify that prompts are called correctly with the right message. 

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/2217)

### CODE REVIEW

**Comment by yao-msft** [comment]
This change will probably block com callers with packages having warnings. We need to expose something in InstallOptions in Com interface

File: src/AppInstallerCLICore/Argument.cpp [hunk 1]
```diff
@@ -97,6 +97,8 @@ namespace AppInstaller::CLI
             return Argument{ "wait", NoAlias, Args::Type::Wait, Resource::String::WaitArgumentDescription, ArgumentType::Flag, false };
         case Args::Type::ProductCode:
             return Argument{ "product-code", NoAlias, Args::Type::ProductCode, Resource::String::ProductCodeArgumentDescription, ArgumentType::Standard, false };
+        case Args::Type::IgnoreWarnings:
+            return Argument{ "ignore-warnings", NoAlias, Args::Type::IgnoreWarnings, Resource::String::IgnoreWarningsArgumentDescription, ArgumentType::Flag, false };
```
Comment (yao-msft) [inline src/AppInstallerCLICore/Argument.cpp:101]: nit: I think "ignore-install-warnings" would be more accurate, "ignore-warnings" seems too generic. Also please update the variable name to leave room for future warnings

File: src/AppInstallerCLICore/Workflows/InstallFlow.cpp [hunk 2]
```diff
@@ -443,6 +458,7 @@ namespace AppInstaller::CLI::Workflow
     void InstallSinglePackage(Execution::Context& context)
     {
         context <<
+            Workflow::DisplayInstallWarnings <<
```
Comment (yao-msft) [inline src/AppInstallerCLICore/Workflows/InstallFlow.cpp:461]: Need to update InstallMultiplePackages as well for import, dependencies, upgrade all etc

File: src/AppInstallerCLICore/Argument.cpp (hunk 1 above)
Comment (yao-msft) [reply src/AppInstallerCLICore/Argument.cpp:101]: > nit: I think "ignore-install-warnings" would be more accurate, "ignore-warnings" seems too generic. Also please update the variable name to leave room for future warnings

Also you'll need to expose them in each applicable command.

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw [hunk 3]
```diff
@@ -1349,4 +1349,10 @@ Please specify one of them using the `--source` option to proceed.</value>
   <data name="ShowLabelInstallationNotes" xml:space="preserve">
     <value>InstallationNotes:</value>
   </data>
+  <data name="IgnoreWarningsArgumentDescription" xml:space="preserve">
+    <value>Disables warning prompts</value>
```
Comment (yao-msft) [inline src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw:1353]: > <value>Disables warning prompts</value> [](http://example.com/codeflow?start=3&length=40)

nit: Disables install warning prompts

**Review by yao-msft — CHANGES_REQUESTED** [review]
:clock1:

**Comment by yao-msft** [comment]
Since this field is per installer level, maybe for better user experience, shall we change the installer selection logic to pick installers without warnings over warnings?

File: src/AppInstallerCLICore/Argument.cpp (hunk 1 above)
Comment (denelon) [reply src/AppInstallerCLICore/Argument.cpp:101]: I've seen a couple of features reported with similar concerns like:
* #715

I just added it to:
* #2128

**Comment by Trenly** [comment]
> Since this field is per installer level, maybe for better user experience, shall we change the installer selection logic to pick installers without warnings over warnings?

I would say no - I wouldn't want winget to select an x86 installer for a x64 system if the x64 installer had installer warnings. Similarly, I wouldn't want the warnings to change a user scoped installation to a machine scoped installation, just because the machine scope didn't have warnings. I could see adding a setting to always suppress the warnings, but I don't think that changing the selection behavior based on a warning is a good idea, since it isn't a guarantee that the application is running when the user attempts the upgrade

**Comment by yao-msft** [comment]
> > Since this field is per installer level, maybe for better user experience, shall we change the installer selection logic to pick installers without warnings over warnings?
> 
> I would say no - I wouldn't want winget to select an x86 installer for a x64 system if the x64 installer had installer warnings. Similarly, I wouldn't want the warnings to change a user scoped installation to a machine scoped installation, just because the machine scope didn't have warnings. I could see adding a setting to always suppress the warnings, but I don't think that changing the selection behavior based on a warning is a good idea, since it isn't a guarantee that the application is running when the user attempts the upgrade

I agree, the installer selection logic actually has different levels of priority when determining which installer to pick, the priority of the InstallWarnings should always be in the lowest. Only when 2 installers have same scope, arch etc would the InstallWarnings be considered a factor. But given how rare this would be, I'm ok with not implementing anything. Just throwing some thought out.

File: src/AppInstallerCLICore/Workflows/InstallFlow.cpp [hunk 4]
```diff
@@ -158,6 +158,21 @@ namespace AppInstaller::CLI::Workflow
         }
     }
 
+    void DisplayInstallWarnings(Execution::Context& context)
+    {
+        auto displayInstallWarnings = context.Get<Execution::Data::Installer>().value().DisplayInstallWarnings;
+
+        if (displayInstallWarnings && !context.Args.Contains(Execution::Args::Type::IgnoreWarnings) && !Settings::User().Get<Settings::Setting::InstallIgnoreWarnings>())
+        {
+            bool ignoreWarning = context.Reporter.PromptForBoolResponse(Resource::String::InstallWarning);
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Workflows/InstallFlow.cpp:167]: I'm not sure what the use case is here, but I don't like prompting on a warning.  At a minimum I would think that the setting should be ignore/display/prompt with the default being display.

ignore = Don't show them at all
display = Display the warning once the installer is chosen (before download) which gives an interactive user a fair amount of time to react with CTRL+C
prompt = This behavior of prompting after displaying the warnings.

File: src/AppInstallerCLICore/Argument.cpp (hunk 1 above)
Comment (JohnMcPMS) [reply src/AppInstallerCLICore/Argument.cpp:101]: Breaking or not, we should change `--force` per this discussion: #2112

**Review by JohnMcPMS — CHANGES_REQUESTED** [review]
I don't think that this feature has been though through enough.  I would expect that many upgrades affect their own package, which is in line with the current message of "interfering with running applications".  So the flag is added, then nearly every package prompts you?

I get that it is painful if running an installer messes with active processes, but I don't think this is the right way to deal with that.

File: src/AppInstallerCLICore/Workflows/InstallFlow.cpp (hunk 4 above)
Comment (Trenly) [reply src/AppInstallerCLICore/Workflows/InstallFlow.cpp:167]: I agree that not prompting is better, especially knowing that WinGet is used in many autonomous install/upgrade scenarios

File: src/AppInstallerCLICore/Workflows/InstallFlow.cpp (hunk 2 above)
Comment (ryfu-msft) [reply src/AppInstallerCLICore/Workflows/InstallFlow.cpp:470]: Added to `InstallMultiplePackages`

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw (hunk 3 above)
Comment (ryfu-msft) [reply src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw]: Changed to `Disables install warnings`

**Pull request closed without merging**